File: src/bmsx_cpp/machine/devices/input/query_port.cpp
#include "machine/devices/input/query_port.h"

#include "machine/bus/io.h"
#include "machine/cpu/string_pool.h"
#include "machine/devices/input/registers.h"

namespace bmsx {

InputControllerQueryPort::InputControllerQueryPort(Memory& memory, const StringPool& strings, InputControllerRegisterFile& registers, InputControllerActionTable& actionTable)
	: m_memory(memory)
	, m_strings(strings)
	, m_registers(registers)
	, m_actionTable(actionTable) {
}

// disable-next-line single_line_method_pattern -- memory-map callbacks require a C-style thunk into the input query-port instance.
void InputControllerQueryPort::writeQueryThunk(void* context, u32, Value value) {
	static_cast<InputControllerQueryPort*>(context)->writeQuery(value);
}

// disable-next-line single_line_method_pattern -- memory-map callbacks require a C-style thunk into the input query-port instance.
void InputControllerQueryPort::writeConsumeThunk(void* context, u32, Value value) {
	static_cast<InputControllerQueryPort*>(context)->writeConsume(value);
}

void InputControllerQueryPort::writeQuery(Value value) {
	m_registers.write(IO_INP_QUERY, value);
	const std::string& queryText = m_strings.toString(m_registers.state.queryStringId);
	m_actionTable.queryAction(static_cast<i32>(m_registers.state.player), queryText, m_queryResult);
	m_registers.writeResult(m_memory, m_queryResult.statusWord, m_queryResult.valueQ16);
}

void InputControllerQueryPort::writeConsume(Value value) {
	m_registers.write(IO_INP_CONSUME, value);
	const std::string& actionNames = m_strings.toString(m_registers.state.consumeStringId);
	m_actionTable.consumeActions(static_cast<i32>(m_registers.state.player), actionNames);
}

} // namespace bmsx

File: src/bmsx/machine/devices/input/query_port.ts
import { Memory } from '../../memory/memory';
import type { Value } from '../../cpu/cpu';
import type { StringPool } from '../../cpu/string_pool';
import {
	IO_INP_CONSUME,
	IO_INP_QUERY,
} from '../../bus/io';
import { InputControllerActionTable, type InputControllerQueryResult } from './action_table';
import { InputControllerRegisterFile } from './registers';

export class InputControllerQueryPort {
	private readonly queryResult: InputControllerQueryResult = { statusWord: 0, valueQ16: 0 };

	public constructor(
		private readonly memory: Memory,
		private readonly strings: StringPool,
		private readonly registers: InputControllerRegisterFile,
		private readonly actionTable: InputControllerActionTable,
	) {}

	public writeQuery(_addr: number, value: Value): void {
		this.registers.write(IO_INP_QUERY, value);
		const queryText = this.strings.toString(this.registers.state.queryStringId);
		this.actionTable.queryAction(this.registers.selectedPlayerIndex(), queryText, this.queryResult);
		this.registers.writeResult(this.memory, this.queryResult.statusWord, this.queryResult.valueQ16);
	}

	public writeConsume(_addr: number, value: Value): void {
		this.registers.write(IO_INP_CONSUME, value);
		const actionNames = this.strings.toString(this.registers.state.consumeStringId);
		this.actionTable.consumeActions(this.registers.selectedPlayerIndex(), actionNames);
	}
}
